python/ola/PidStore.py

```diff
@@ -157,14 +157,58 @@ def ValidateAddressing(self, args, command_class):
         return False
     return True
 
-  def __cmp__(self, other):
-    return cmp(self._value, other._value)
+  def _GroupCmp(self, a, b):
+    def setToDescription(x):
+      def descOrNone(y):
+        return y.GetDescription() if y is not None else "None"
+      return ':'.join([descOrNone(x[f])
+                       for f in [RDM_GET, RDM_SET, RDM_DISCOVERY]])
+    ax = setToDescription(a)
+    bx = setToDescription(b)
+    return (ax > bx) - (ax < bx)  # canonical for cmp(a,b)
+
+  def __eq__(self, other):
+    if not isinstance(other, self.__class__):
+      return False
+    return (self.value == other.value and
+            self.name == other.name and
+            self._GroupCmp(self._requests, other._requests) == 0 and
+            self._GroupCmp(self._responses, other._responses) == 0)
+
+  def __lt__(self, other):
+    if not isinstance(other, self.__class__):
+      return NotImplemented
+    if self.value != other.value:
+      return self._value < other._value
+    if self.name != other.name:
+      return self.name < other.name
+    return (self._GroupCmp(self._requests, other._requests) < 0 or
+            self._GroupCmp(self._responses, other._responses) < 0)
+
+  # These 4 can be replaced with functools:total_ordering when 2.6 is dropped
+  def __le__(self, other):
+    if not isinstance(other, self.__class__):
+      return NotImplemented
+    return self < other or self == other
+
+  def __gt__(self, other):
+    if not isinstance(other, self.__class__):
+      return NotImplemented
+    return not self <= other
+
+  def __ge__(self, other):
+    if not isinstance(other, self.__class__):
+      return NotImplemented
+    return not self < other
+
+  def __ne__(self, other):
+    return not self == other
 
   def __str__(self):
     return '%s (0x%04hx)' % (self.name, self.value)
 
   def __hash__(self):
-    return self._value
+    return hash((self._name, self._value))
```

Comment:
I think value is the key bit (and manufacturer where present, but I think that's handled in a different way here). I think if we have two PIDs with the same numeric value and different names, we probably WANT to have a collision, or flag an exception or whatever.

Reply:
So I honestly haven't learned enough about the protocol to have opinions about those aspects.  In truth this is a bit weaker than it ought to be, as a == b => hash(a) == hash(b) and this only tests name and value (wimping out b/c traversing requests etc would be a lot of code) whereas eq uses GroupCmp to compare requests and responses.  I thought that was unlikely to cause a problem, though.  I can change it to something else, but IMHO hash and eq/lt should match behavior here.

My understanding is the loader actually enforces not loading duplicate PIDs unless override is set, so in practice it shouldn't be a problem.

Reply:
So broadly the value is the parameter ID sent on the wire, the name is just a friendly label we display for it within the software, so if they're both 0x0042, but one is called foo and one bar, the actual data on the wire and response from the responder will be the same, just how we display part of it will change, hence wanting a collision.

Reply:
There's that other conversation I opened on the matching eq/lt where Nils added the other fields beyond value.  It sounds like you would rather eq, lt, and hash all consider only value?

Reply:
I'm going to make that change in the next changeset.

Reply:
Name should definitely be dropped, request/response is TBC a little bit of testing of the C++ and Protobuf behaviour.

Reply:
As mentioned elsewhere this is just id which you've done.